aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h

```diff
@@ -105,11 +106,91 @@ struct with_scattered_tensor_options_<FuncPtr, guts::typelist::typelist<Paramete
     }
 };
 
+/**
+ * Take a kernel function that has a `TensorOptions` argument and
+ * return a new kernel function that has `optional<ScalarType>,
+ * optional<Layout>, optional<Device>, optional<bool>` arguments
+ * instead, packs them into a `TensorOptions` struct and
+ * calls the original kernel function.
+ */
+template<class FuncPtr>
+constexpr auto with_scattered_tensor_options(FuncPtr) {
+    return typename with_scattered_tensor_options_<FuncPtr>::FuncPtr();
 }
 
-template<class FuncPtr>
-constexpr auto hacky_wrapper_for_legacy_signatures(FuncPtr) {
-    return typename detail::with_scattered_tensor_options<FuncPtr>::FuncPtr();
+template<class KernelType>
+struct make_optional_tensor_explicit final {
+    template<class T>
+    static decltype(auto) call(T&& arg) {
+        return std::forward<T>(arg);
+    }
+};
+
+template<>
+struct make_optional_tensor_explicit<at::Tensor> final {
+    template<class _Tensor, std::enable_if_t<std::is_same<std::remove_cv_t<std::remove_reference_t<_Tensor>>, at::Tensor>::value, int> = 0>
+    static decltype(auto) call(_Tensor&& arg) {
+        return std::forward<_Tensor>(arg);
+    }
+
+    template<class _OptTensor, std::enable_if_t<std::is_same<std::remove_cv_t<std::remove_reference_t<_OptTensor>>, optional<at::Tensor>>::value, int> = 0>
+    static at::Tensor call(_OptTensor&& arg) {
+        if (arg.has_value()) {
+            return *std::forward<_OptTensor>(arg);
+        } else {
+            return at::Tensor(); // undefined tensor
+        }
+    }
+};
+
+template<class TargetSignature, class KernelSignature, class KernelFunc>
+struct with_explicit_optional_tensors_ final {};
+
+template<class Return, class... TargetSignatureArgs, class... KernelSignatureArgs, Return(*KernelFunc)(KernelSignatureArgs...)>
+struct with_explicit_optional_tensors_<Return (TargetSignatureArgs...), Return(KernelSignatureArgs...), TORCH_FN_TYPE(KernelFunc)> final {
+    static Return wrapper(TargetSignatureArgs... args) {
+        return (*KernelFunc)(make_optional_tensor_explicit<std::remove_cv_t<std::remove_reference_t<KernelSignatureArgs>>>::call(
+                std::forward<TargetSignatureArgs>(args)
+            )...);
+    }
+};
+
+template<class T>
+constexpr bool _is_optional_tensor_arg() {
+    return std::is_same<c10::optional<at::Tensor>, std::decay_t<T>>::value;
+}
+template<class T> using is_optional_tensor_arg = guts::bool_constant<_is_optional_tensor_arg<T>()>;
+
+/**
+ * Take a kernel function that has a number of `Tensor` arguments
+ * and take in a `TargetSignature` that must match, but is allowed
+ * to take `optional<Tensor>` in place of some or all of the `Tensor`
+ * arguments. Returns a new kernel function that has `optional<Tensor>`
```

Comment:
Personally I'd really benefit from comments on each template definition describing what it does, in addition to a summary comment like this on the public endpoint. E.g. here, everything that's being described (other than the optional Tensor detection) is happening somewhere else, in uncommented templates. Doesn't have to be in depth, just something to label the pieces basically.

Reply:
I added some more comments